oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,347 @@
+
+    // Append the canonical headers.
+    StringBuilder canonicalHeaders = new StringBuilder();
+    for (String headerName : sortedHeaderNames) {
+      canonicalHeaders.append(headerName + ":" + headers.get(headerName)).append(NEW_LINE);
+    }
+    sb.append(canonicalHeaders).append(NEW_LINE);
+
+    // Append the signed headers.
+    sb.append(Joiner.on(';').join(sortedHeaderNames)).append(NEW_LINE);
+
+    // Append the hashed request payload.
+    sb.append(DigestUtils.sha256Hex(requestPayload));
+
+    // Return the hashed canonical request.
+    return DigestUtils.sha256Hex(sb.toString());
+  }
+
+  /** Task 2: Create a string to sign for Signature Version 4. */
+  private String createStringToSign(
+      String canonicalRequestHash, String xAmzDate, String credentialScope) {
+    return HASHING_ALGORITHM
+        + NEW_LINE
+        + xAmzDate
+        + NEW_LINE
+        + credentialScope
+        + NEW_LINE
+        + canonicalRequestHash;
+  }
+
+  /**
+   * Task 3: Calculate the signature for AWS Signature Version 4.
+   *
+   * @param date The date used in the hashing process in YYYYMMDD format.
+   */
+  private String calculateAwsV4Signature(
+      String serviceName, String secret, String date, String region, String stringToSign) {
+    byte[] kDate = sign(("AWS4" + secret).getBytes(UTF_8), date.getBytes(UTF_8));
+    byte[] kRegion = sign(kDate, region.getBytes(UTF_8));
+    byte[] kService = sign(kRegion, serviceName.getBytes(UTF_8));
+    byte[] kSigning = sign(kService, AWS_REQUEST_TYPE.getBytes(UTF_8));
+    return BaseEncoding.base16().encode(sign(kSigning, stringToSign.getBytes(UTF_8))).toLowerCase();
+  }
+
+  private Map<String, String> getCanonicalHeaders(String date) {
+    Map<String, String> headers = new HashMap<>();
+    headers.put("host", uri.getHost());
+
+    // Only add the date if it hasn't been specified through the "date" header.
+    if (!additionalHeaders.containsKey("date")) {
+      headers.put("x-amz-date", date);
+    }
+
+    if (awsSecurityCredentials.getToken() != null && !awsSecurityCredentials.getToken().isEmpty()) {
+      headers.put("x-amz-security-token", awsSecurityCredentials.getToken());
+    }
+
+    // Add all additional headers.
+    for (String key : additionalHeaders.keySet()) {
+      // Header keys need to be lowercase.
+      headers.put(key.toLowerCase(), additionalHeaders.get(key));
+    }
+    return headers;
+  }
+
+  private AwsDates getDates() {
+    DateFormat dateFormat = new SimpleDateFormat(AWS_DATE_FORMAT);
+    dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+
+    String xAmzDate = null;
+    if (additionalHeaders.containsKey("x-amz-date")) {
+      xAmzDate = additionalHeaders.get("x-amz-date");
+    }
+
+    if (additionalHeaders.containsKey("date")) {
+      String date = additionalHeaders.get("date");
+      Date inputDate;
+      try {
+        inputDate = new SimpleDateFormat("E, dd MMM yyyy HH:mm:ss z").parse(date);
+      } catch (ParseException e) {
```

Comment:
Make the try block cover more lines so the variable declaration and initialization aren't split